REGRESSION (264004@main): Images MotionMark subtest shows incorrect rendering on Intel

kkinnunen-apple opened the pull request:
#### de645b364760aec6f0f19234d97392020340d1bf
<pre>
REGRESSION (264004@main): Images MotionMark subtest shows incorrect rendering on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=258560">https://bugs.webkit.org/show_bug.cgi?id=258560</a>
rdar://111197581

Reviewed by Simon Fraser.

ImageBufferShareableMappedIOSurfaceBitmapBackend would lock the
IOSurface and draw to it as a bitmap.

The code was written with the expectation was that after layer backing
store flush, ImageBuffer::releaseGraphicsContext() was called and this
would release the GraphicsContext, the underlying CGContext and the
IOSurface lock. This was a misunderstanding, releaseGraphicsContext() is
never called. This would cause MotionMark Images test to render
incorrectly on Intel-based Macs, since the IOSurface unlock is needed to
copy the main memory IOSurface cache back to the real IOSurface memory.

After each layer backing store update, the corresponding layer
ImageBuffers are flushed. Use the flush as the cue to unlock the
IOSurface that ImageBufferShareableMappedIOSurfaceBitmapBackend draws
to, before external access via the UI process starts.

Related but not strictly needed, make sure IOSurface locking errors are
handled gracefully. This bug is not about IOSurface locking errors, but
they were ignored before, and as such the behavior was part of the
investigation of this bug.

* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
(WebCore::IOSurface::lock):
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::createBitmapPlatformContext):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::~ImageBufferShareableMappedIOSurfaceBitmapBackend):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::context):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::flushContext):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: <a href="https://commits.webkit.org/265712@main">https://commits.webkit.org/265712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e54ee9cad5bb4ed27d6165d1329c98edab81f423

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11535 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12099 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13180 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10995 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11556 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14125 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13870 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12562 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9781 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13602 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9854 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10474 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17617 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10628 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13810 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9084 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10199 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2814 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14474 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10877 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/cocoa/IOSurface.h:
```diff
@@ -123,8 +115,15 @@ class IOSurface final {
 
     private:
+        Locker(IOSurfaceRef surface, uint32_t flags)
+            : m_surface(surface)
```
It's a bit surprising that this class doesn't retain the surface.

kkinnunen-apple replied:
Yeah. It can go either way. I suppose having a ref is more in line with the expectations. Added.
(It's a bit unfortunate that storing Locker duplicates the storage for the surface.)

smfr commented on Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
```diff
@@ -74,15 +74,41 @@ ImageBufferBackendHandle ImageBufferShareableMappedIOSurfaceBitmapBackend::creat
+            return *m_context;
+        }
     }
+    // For some reason we ran into an error. Construct an invalid context, with current API we must
```
Should we RELEASE_LOG here?

kkinnunen-apple replied:
I suppose. I'm not a fan but maybe that motivates handling the errors properly in the caller (and then moving the release log there, and then repeating the procedure..)

smfr commented on Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
```diff
@@ -74,15 +74,41 @@ ImageBufferBackendHandle ImageBufferShareableMappedIOSurfaceBitmapBackend::creat
+            auto bitsPerPixel = bitsPerComponent * 4;
+            auto bytesPerRow = m_surface->bytesPerRow();
+            auto size = m_surface->size();
+            auto success = CGBitmapContextSetData(cgContext, 0, 0, size.width(), size.height(), lock->surfaceBaseAddress(), bitsPerComponent, bitsPerPixel, bytesPerRow);
```
Is there a data copy under here, or is this just updating the context with a new base address?

kkinnunen-apple replied:
Updating the context with a new base address.

kkinnunen-apple committed:
REGRESSION (264004@main): Images MotionMark subtest shows incorrect rendering on Intel
https://bugs.webkit.org/show_bug.cgi?id=258560
rdar://111197581

Reviewed by Simon Fraser.

ImageBufferShareableMappedIOSurfaceBitmapBackend would lock the
IOSurface and draw to it as a bitmap.

The code was written with the expectation was that after layer backing
store flush, ImageBuffer::releaseGraphicsContext() was called and this
would release the GraphicsContext, the underlying CGContext and the
IOSurface lock. This was a misunderstanding, releaseGraphicsContext() is
never called. This would cause MotionMark Images test to render
incorrectly on Intel-based Macs, since the IOSurface unlock is needed to
copy the main memory IOSurface cache back to the real IOSurface memory.

After each layer backing store update, the corresponding layer
ImageBuffers are flushed. Use the flush as the cue to unlock the
IOSurface that ImageBufferShareableMappedIOSurfaceBitmapBackend draws
to, before external access via the UI process starts.

Related but not strictly needed, make sure IOSurface locking errors are
handled gracefully. This bug is not about IOSurface locking errors, but
they were ignored before, and as such the behavior was part of the
investigation of this bug.

* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::getPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
(WebCore::IOSurface::lock):
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::createBitmapPlatformContext):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::~ImageBufferShareableMappedIOSurfaceBitmapBackend):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::context):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::flushContext):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: https://commits.webkit.org/265712@main
